### PR TITLE
SNOW-543664 offset token retries + Refactor to reopen channel on SFException 

### PR DESCRIPTION
**Pull request by sfc-gh-japatel**
Refactor 1:
- Move openChannel inside TopicPartitionChannel since it can be used to re-open in case of failures. 
- Introducing a new library which implements retries, fallback and circuitbreaker. 
  - https://github.com/failsafe-lib/failsafe
  - I decided to not implement this because this can also be extended to various other retry mechanisms such as exponential and jitters. 
  - Also, this library is very widely used. 
- Added tests to verify the behavior of retries. 
  - TLDR: Only SFException will be retried and fallback to opening the channel. 
  - No other exception will result into retry.

### CODE REVIEW

**Comment by sfc-gh-tzhang** [comment]
> Introducing a new library which implements retries, fallback and circuitbreaker.
> * https://github.com/failsafe-lib/failsafe
> * I decided to not implement this because this can also be extended to various other retry mechanisms such as exponential and jitters.
> * Also, this library is very widely used.

Do we need any security approval? (assuming no other SF connector/SDK is using this)

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 1]
```diff
@@ -323,37 +366,141 @@ public long getOffsetSafeToCommitToKafka() {
...
+   *
+   * @return the fallback function to execute when all retries from getOffsetToken have exhausted.
+   */
+  private Fallback<Long> getFallbackForGetOffsetTokenFailure() {
+    return Fallback.builder(
+            () -> {
+              openChannelForTable();
+              return fetchLatestCommittedOffsetFromSnowflake();
+            })
+        .handle(SFException.class)
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:436]: what will we do for other type of exceptions?

**Comment by sfc-gh-japatel** [comment]
> > Introducing a new library which implements retries, fallback and circuitbreaker.
> > 
> > * https://github.com/failsafe-lib/failsafe
> > * I decided to not implement this because this can also be extended to various other retry mechanisms such as exponential and jitters.
> > * Also, this library is very widely used.
> 
> Do we need any security approval? (assuming no other SF connector/SDK is using this)

Yes, we got the approval. https://snowflakecomputing.atlassian.net/browse/CASEC-3491

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 2]
```diff
@@ -323,37 +366,141 @@ public long getOffsetSafeToCommitToKafka() {
...
+  private SnowflakeStreamingIngestChannel openChannelForTable() {
+    OpenChannelRequest channelRequest =
+        OpenChannelRequest.builder(channelName)
+            .setDBName(this.databaseName)
+            .setSchemaName(this.schemaName)
+            .setTableName(this.tableName)
+            .setOnErrorOption(OpenChannelRequest.OnErrorOption.CONTINUE)
+            .build();
+    LOGGER.info(
+        "Opening a channel with name:{} for table name:{}", this.channelName, this.tableName);
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:500]: nit: is it possible we distinguish between open and reopen for better debugging purpose?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 3]
```diff
@@ -146,12 +144,16 @@ public void startTask(String tableName, String topic, int partition) {
...
-        new TopicPartitionChannel(partitionChannel, this.kafkaRecordErrorReporter);
-    partitionsToChannel.put(partitionChannelKey, topicPartitionChannel);
+    // Create new instance of TopicPartitionChannel which will always open the channel.
+    partitionsToChannel.put(
+        partitionChannelKey,
+        new TopicPartitionChannel(
+            this.streamingIngestClient,
+            partitionChannelKey,
+            this.connectorConfig.get(Utils.SF_DATABASE),
+            this.connectorConfig.get(Utils.SF_DATABASE),
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:154]: yess.. SF_SCHEMA. thanks for the catch!

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 4]
```diff
@@ -323,37 +366,141 @@ public long getOffsetSafeToCommitToKafka() {
...
+    return Fallback.builder(
+            () -> {
+              openChannelForTable();
+              return fetchLatestCommittedOffsetFromSnowflake();
+            })
+        .handle(SFException.class)
+        .onFailure(
+            event ->
+                LOGGER.warn(
+                    "[FALLBACK] Failed to open Channel/fetch offsetToken for channel:{}",
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:440]: I was planning to use it just for fallback in offsettoken, but will keep it consistent. thanks

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 2 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:500]: Yes.. I actually add more debugging statements from reopening the channel.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 5]
```diff
@@ -91,11 +111,29 @@
...
+   */
   public TopicPartitionChannel(
-      SnowflakeStreamingIngestChannel channel, KafkaRecordErrorReporter kafkaRecordErrorReporter) {
-    this.channel = channel;
-    this.kafkaRecordErrorReporter = kafkaRecordErrorReporter;
+      SnowflakeStreamingIngestClient streamingIngestClient,
+      final String channelName,
+      final String databaseName,
+      final String schemaName,
+      final String tableName,
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:127]: We will need to use this while opening the channel.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 1 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:436]: Only SFExceptions are being retried. For runtime exceptions, it will throw up and surface back same errors.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 6]
```diff
@@ -323,37 +366,141 @@ public long getOffsetSafeToCommitToKafka() {
...
+   * after reopen is successful.
+   *
+   * <p>Fallback is only attempted on SFException
+   *
+   * @return the fallback function to execute when all retries from getOffsetToken have exhausted.
+   */
+  private Fallback<Long> getFallbackForGetOffsetTokenFailure() {
+    return Fallback.builder(
+            () -> {
+              openChannelForTable();
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:433]: This is incorrect actually, I will need to assign the this.channel with object returned from openChannelForTable. 
I will fix this.

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:433]: Fixed in revision.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:440]: I have also added few more log statements for debugging but in another PR. this is solely for retrying..

**Comment by sfc-gh-japatel** [comment]
I actually have added IT test in upcoming PR which creates a channel outside of SinkServiceV2 and then the service tries to insert the record which gets SFException. (More on that in later PR)

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 1 above)
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:439]: This is fine for now, but we may want to retry for some of the runtime exceptions as well

**Comment by sfc-gh-japatel** [comment]
> I actually have added IT test in upcoming PR which creates a channel outside of SinkServiceV2 and then the service tries to insert the record which gets SFException. (More on that in later PR)

[src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelIT.java](https://github.com/snowflakedb/snowflake-kafka-connector/pull/400/files#diff-e378884af1a0a01647afdde997a469718c3c6d247dd9cc7a1ee68ee2c431f548)

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM!
